### ISSUE TITLE
Dask Operator becomes unresponsive after ~1hr

### ISSUE THREAD

**Issue by baswelsh** [dask/dask-kubernetes#620]
This one's a bit tricky.  I haven't been able to reproduce it in a kind or minikube cluster - it only happens in our hosted environments (i.e., AKS, EKS, etc.).  The dask operator deployment works great for a while and then becomes unresponsive.  I've seen it happen after being up for 1hr and 2hrs (pretty much on the nose).

The status of the pod is `running` but the logs are frozen and the operator is unresponsive to `KubeCluster` instantiation.  Has anyone seen this?  If not, I'd really appreciate any guidance on how to efficiently narrow in on the root cause (i.e., increasing log levels, inspecting heartbeats, operator health/status queries, etc.)

**Anything else we need to know?**:

Note that resetting the dask operator deployment via `kubectl rollout restart deployment dask-kubernetes-operator -n dask-operator` recovers the operator after about 60+s (which is how long it takes to terminate the pod once frozen).

Here are the final lines of the `dask-kubernetes-operator` pod logs:

```
[2022-11-28 00:45:46,745] kopf.objects         [DEBUG   ] [6556cd3b-4385-4d5d-98dd-af70b0ad8f80/sedaro-dask] Updating diff: (('change', ('status', 'phase'), 'Created', 'Running'),)
[2022-11-28 00:45:46,745] kopf.objects         [DEBUG   ] [6556cd3b-4385-4d5d-98dd-af70b0ad8f80/sedaro-dask] Patching with: <redacted>
[2022-11-28 00:45:46,754] kubernetes_asyncio.c [DEBUG   ] response body: <redacted>
[2022-11-28 00:45:46,764] kubernetes_asyncio.c [DEBUG   ] response body: <redacted>
[2022-11-28 00:45:46,764] kopf.objects         [INFO    ] [6556cd3b-4385-4d5d-98dd-af70b0ad8f80/sedaro-dask-default] Successfully adopted by sedaro-dask
[2022-11-28 00:45:46,784] kubernetes_asyncio.c [DEBUG   ] response body: <redacted>
[2022-11-28 00:45:46,801] kubernetes_asyncio.c [DEBUG   ] response body: <redacted>
[2022-11-28 00:45:46,805] kopf.objects         [INFO    ] [6556cd3b-4385-4d5d-98dd-af70b0ad8f80/sedaro-dask-default] Scaled worker group sedaro-dask-default up to 1 workers.
[2022-11-28 00:45:46,805] kopf.objects         [INFO    ] [6556cd3b-4385-4d5d-98dd-af70b0ad8f80/sedaro-dask-default] Handler 'daskworkergroup_create' succeeded.
[2022-11-28 00:45:46,806] kopf.objects         [INFO    ] [6556cd3b-4385-4d5d-98dd-af70b0ad8f80/sedaro-dask-default] Creation is processed: 1 succeeded; 0 failed.
[2022-11-28 00:45:46,806] kopf.objects         [DEBUG   ] [6556cd3b-4385-4d5d-98dd-af70b0ad8f80/sedaro-dask-default] Patching with: <redacted>
[2022-11-28 00:45:46,858] kopf.objects         [DEBUG   ] [6556cd3b-4385-4d5d-98dd-af70b0ad8f80/sedaro-dask] Something has changed, but we are not interested (the essence is the same).
[2022-11-28 00:45:46,858] kopf.objects         [DEBUG   ] [6556cd3b-4385-4d5d-98dd-af70b0ad8f80/sedaro-dask] Handling cycle is finished, waiting for new changes.
[2022-11-28 00:45:46,927] kopf.objects         [DEBUG   ] [6556cd3b-4385-4d5d-98dd-af70b0ad8f80/sedaro-dask-default] Something has changed, but we are not interested (the essence is the same).
[2022-11-28 00:45:46,927] kopf.objects         [DEBUG   ] [6556cd3b-4385-4d5d-98dd-af70b0ad8f80/sedaro-dask-default] Handling cycle is finished, waiting for new changes.
[2022-11-28 00:46:32,107] kopf.objects         [DEBUG   ] [6556cd3b-4385-4d5d-98dd-af70b0ad8f80/sedaro-dask] Deleted, really deleted, and we are notified.
[2022-11-28 00:46:32,152] kopf.objects         [DEBUG   ] [6556cd3b-4385-4d5d-98dd-af70b0ad8f80/sedaro-dask-default] Deleted, really deleted, and we are notified.
```

**Environment**:

- Dask version: 2022.10.1
- Python version: 3.9.15
- Operating System: Linux
- Install method (conda, pip, source): pip

**Comment by baswelsh** [comment]
Here is the specific `KubeCluster` block:

```python
cluster = KubeCluster(
    namespace=namespace,
    name=name,
    image=image,
    worker_command=f'dask-worker --nworkers={CAP}',
    n_workers=numWorkers,
    shutdown_on_close=False,
    env={
        'PYTHONPATH': '/app',
    },
    resources={"worker": {"requests": {"cpu": CAP}, "limits": {"cpu": CAP}}},
)
```

**Comment by jacobtomlinson** [comment]
@baswelsh I've just tagged `2022.11.1` which has the liveness probe. Could you give that a try? It would be interesting to see if it starts failing and restarts the pod (which we should still find the root cause of).

**Comment by baswelsh** [comment]
Thanks @jacobtomlinson.  I deployed the latest operator via helm and an initial test seems to indicate that it is fixed - although I am running a few more tests throughout the day.  Note that prior to this, I was installing the operator via the old method of using manifests.  Could this have been the cause of the original issue?  

I haven't seen the pod restart (due to poor health or otherwise) so I'm not sure the liveliness probe is the fix in this case.  But as you mentioned, it is great to now have it.

Let me know what you think about the manifest installation being the issue.  I will report back regarding the results of the follow up testing.  If I can't recreate the issue over the next ~8hrs, we can close this 👍

**Comment by baswelsh** [comment]
Scratch that - a little longer and it locked up.  I don't see an events under `kubectl describe pod/deployment`.  Where could I look to see if the health checks are running?  The pod was never replaced on unhealthy.

**Comment by baswelsh** [comment]
Further findings:

Once locked up, I am able to hit the `/healthz` route from a different pod and receive responses indicative of good dask-operator pod health.

```bash
$ curl http://10.0.10.18:8080/healthz
{"now": "2022-11-28T21:33:18.197972"}
$ curl http://10.0.10.18:8080/healthz
{"now": "2022-11-28T21:33:18.197972"}
$ curl http://10.0.10.18:8080/healthz
{"now": "2022-11-28T21:33:18.197972"}
$ curl http://10.0.10.18:8080/healthz
{"now": "2022-11-28T21:33:34.375675"}
$ curl http://10.0.10.18:8080/healthz
{"now": "2022-11-28T21:33:34.375675"}
```

In my scenario, I have one scheduler and one worker.  I have now seen it lock up twice in the following way - It creates the scheduler and the Worker Group but seizes before it can scale the Worker Group up to one Worker from zero. This isn't consistent, sometimes I don't get a scheduler either.  But it is interesting as it possibly indicates that requests trigger it instead of dormant periods.

**Comment by jacobtomlinson** [comment]
I very much appreciate the rolling commentary, I don't think I'll be able to get to the bottom of this without your debugging help.

It makes sense that the replacement controller catches up. The kopf controller is designed to be interruptable (like if the node it is running on gets replaced) and I think it uses annotations to track what has and hasn't been operated on.

I was hoping that the health checks would also lock up and the controller would get restarted but I guess I was wrong (maybe the health check runs in a new thread or something). It's not a fix but still a hacky workaround that would be useful. So yeah getting the health endpoint to know that it is deadlocked would be useful, I'm not really sure how to make that happen though.

I wonder if the event loop is getting blocked by something. There is one `while True` loop which is the only obvious place that it could be getting stuck. There is already #614 that removes this so maybe I should try and get that PR over the line.

Do any of your schedulers use a loadbalancer?

**Comment by jacobtomlinson** [comment]
I spotted https://github.com/nolar/kopf/issues/847 which looks like a similar problem around kopf deadlocking on AKS after some idle period. 

That issue suggests configuring some timeouts to work around this. I remembered we have #486 open which includes similar changes suggested by @philipp-sontag-by for timeout reasons. That PR is failing for reasons I still don't understand so I've raised #626 with the relevant subset which may resolve this.

**Comment by philipp-sontag-by** [comment]
Yes we had similar problems with our kopf operator. We haven't seen them since we configured the timeout settings though
```python
    # Set server and client timeouts to reconnect from time to time.
    # In rare occasions the connection might go idle we will no longer receive any events.
    # These timeouts should help in those cases.
    # https://github.com/nolar/kopf/issues/698
    # https://github.com/nolar/kopf/issues/204
    settings.watching.server_timeout = 120
    settings.watching.client_timeout = 150
    settings.watching.connect_timeout = 5

```

**Comment by baswelsh** [comment]
Thank you both!  I have a load balancer for pod/node egress out of AKS. 

Is there a way to try these settings in the 2022.11.1 helm release?  Can they be configured after the fact?

**Comment by jacobtomlinson** [comment]
@baswelsh once #626 passes CI I'll merge it and release `2022.11.2` so you can try it out.

**Comment by jacobtomlinson** [comment]
Oops didn't mean to close. I've tagged `2022.11.2` which should be available on PyPI, Conda Forge, GitHub container registry, Helm repo, etc shortly. @baswelsh could you give it a test and let us know how it goes?

**Comment by baswelsh** [comment]
Awesome.  I'll give it a try as soon as they publish and let you know what I find. Thanks @jacobtomlinson!

**Comment by baswelsh** [comment]
This is strange - The container is failing to startup with the following exception:

```
Traceback (most recent call last):
  File "/usr/local/bin/kopf", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/kopf/cli.py", line 59, in wrapper
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/click/decorators.py", line 84, in new_func
    return ctx.invoke(f, obj, *args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/kopf/cli.py", line 109, in run
    loaders.preload(
  File "/usr/local/lib/python3.8/site-packages/kopf/_cogs/helpers/loaders.py", line 46, in preload
    importlib.import_module(name)
  File "/usr/local/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 843, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/usr/local/lib/python3.8/site-packages/dask_kubernetes/operator/controller/__init__.py", line 1, in <module>
    from .controller import *
  File "/usr/local/lib/python3.8/site-packages/dask_kubernetes/operator/controller/controller.py", line 8, in <module>
    import pykube
ModuleNotFoundError: No module named 'pykube'
```

Install command: 

`helm install --wait --repo https://helm.dask.org --create-namespace -n dask-operator --version 2022.11.2 --generate-name dask-kubernetes-operator`

**Comment by baswelsh** [comment]
Looks this be introduced in https://github.com/dask/dask-kubernetes/commit/80c11df1bb80cee307881473b2b0e09d0f447edd

**Comment by baswelsh** [comment]
It looks like `pykube` is installed as a dep to a test dependency which explains why the CI passed. I will open a quick PR @jacobtomlinson.

**Comment by baswelsh** [comment]
Thanks @jacobtomlinson!  So far, so good!  I'm going to close this and will reopen if further testing reproduces it. I appreciate the quick turnaround on this! 💯

**Comment by baswelsh** [comment]
This issue just wont die! This one is my bad @jacobtomlinson - take a look at PR #628 for details.  Without this fix, the latest release is still broken, just in a different way.  Note that the operate has been stable however 🎉 .  This new issue is avoided by using a prior version of the client with the latest operator.

Relevant exception in operator logs:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/kopf/_core/actions/execution.py", line 279, in execute_handler_once
    result = await invoke_handler(
  File "/usr/local/lib/python3.8/site-packages/kopf/_core/actions/execution.py", line 374, in invoke_handler
    result = await invocation.invoke(
  File "/usr/local/lib/python3.8/site-packages/kopf/_core/actions/invocation.py", line 139, in invoke
    await asyncio.shield(future)  # slightly expensive: creates tasks
  File "/usr/local/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.8/site-packages/dask_kubernetes/operator/controller/controller.py", line 284, in handle_scheduler_service_status
    api = pykube.HTTPClient(pykube.KubeConfig.from_env())
AttributeError: type object 'KubeConfig' has no attribute 'from_env'
```

**Comment by jacobtomlinson** [comment]
Thanks for #628. Released `2022.11.4`

**Comment by baswelsh** [comment]
Awesome, thanks @jacobtomlinson.  I will get this tested and closed ASAP!

**Comment by jacobtomlinson** [comment]
It's been a while since there was any activity here so I'm going to close this out. If folks are still running into this feel free to comment here and I'll reopen.
